### PR TITLE
Fix backfill storing incorrect state for events

### PR DESCRIPTION
**Pull request by erikjohnston**
Broadly, we were going through marking all fetched "auth events" as outliers, even if they appeared in the main backfill chunk of events. This led to state getting very confused when we tried to persist events that referenced outliers.

### CODE REVIEW

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -786,13 +801,18 @@ def backfill(self, dest, room_id, limit, extremities):
             })
 
         for e_id in events_to_state:
+            # For paranoia we ensure that these events are marked as
```
Comment (richvdh) [inline synapse/handlers/federation.py]: this seems to be in conflict with the comment at line 775 that says "... state get persisted as outliers". What am I missing?

File: synapse/handlers/federation.py [hunk 2]
```diff
@@ -786,13 +801,18 @@ def backfill(self, dest, room_id, limit, extremities):
...
+            # non-outliers
+            ev = event_map[e_id]
+            ev.internal_metadata.outlier = False
```
Comment (richvdh) [inline synapse/handlers/federation.py]: this sort of thing always alarms me slightly; if `outlier` was set doesn't it mean that something else has gone wrong? shouldn't we just raise an AssertionError or something?

Comment (erikjohnston) [reply synapse/handlers/federation.py]: Assert is actually probably what we want here, yes

File: synapse/handlers/federation.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/handlers/federation.py]: This needs more comments, obviously. The keys of `events_to_state` are the new backwards extremities, as those are the events that we fetched state for

Comment (richvdh) [reply synapse/handlers/federation.py]: it would seem a whole lot less opaque to iterate over `edges` rather than the keys of `events_to_state`, but ok.

**Comment by richvdh** [comment]
[sad pep8 is sad]

File: changelog.d/4718.bugfix [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix paginating over federation persisting incorrect state
```
Comment (richvdh) [inline changelog.d/4718.bugfix]: needs a full-stop

**Pull request merged**